zkvm/src/merkle.rs

```diff
@@ -32,16 +32,32 @@ enum MerkleNode {
 
 impl MerkleTree {
     /// Constructs a new MerkleTree based on the input list of entries.
-    pub fn build<M: MerkleItem>(label: &'static [u8], list: &[M]) -> Option<Self> {
-        if list.len() == 0 {
-            return None;
-        }
-        let t = Transcript::new(label);
-        Some(MerkleTree {
+    pub fn build<M: MerkleItem>(label: &'static [u8], list: &[M]) -> Self {
+        let root = match list.len() {
+            0 => None,
+            _ => {
+                let t = Transcript::new(label);
+                Some(Self::build_tree(t, list))
+            }
+        };
+        MerkleTree {
             size: list.len(),
             label,
-            root: Self::build_tree(t, list),
-        })
+            root: root,
+        }
+    }
+
+    /// Returns the root hash of the Merkle tree.
+    pub fn hash(&self) -> [u8; 32] {
+        match &self.root {
+            None => {
+                let t = Transcript::new(self.label);
+                let mut result = [0u8; 32]; // xxx why must i preinitialize this?
```

Comment:
Because you cannot pass a readable reference to uninitialized data.

Reply:
Is there a Rust idiom that lets me pass a container to a function to be initialized? Like passing a write-iterator or something?

Reply:
There is &mut Vec and there is IntoIterator trait. Or do you specifically mean “i want to avoid memset”? Then you can use unsafe raw memory api in core::mem and core::ptr.